### PR TITLE
[lldb/linux] Make truncated reads work

### PR DESCRIPTION
**Pull request by labath**
Previously, we were returning an error if we couldn't read the whole region. This doesn't matter most of the time, because lldb caches memory reads, and in that process it aligns them to cache line boundaries. As (LLDB) cache lines are smaller than pages, the reads are unlikely to cross page boundaries.

Nonetheless, this can cause a problem for large reads (which bypass the cache), where we're unable to read anything even if just a single byte of the memory is unreadable. This patch fixes the lldb-server to do that, and also changes the linux implementation, to reuse any partial results it got from the process_vm_readv call (to avoid having to re-read everything again using ptrace, only to find that it stopped at the same place).

This matches debugserver behavior. It is also consistent with the gdb remote protocol documentation, but -- notably -- not with actual gdbserver behavior (which returns errors instead of partial results). We filed a
[clarification bug](https://sourceware.org/bugzilla/show_bug.cgi?id=24751) several years ago. Though we did not really reach a conclusion there, I think this is the most logical behavior.

The associated test does not currently pass on windows, because the windows memory read APIs don't support partial reads (I have a WIP patch to work around that).

### CODE REVIEW

File: lldb/test/API/functionalities/memory/holes/TestMemoryHoles.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        self._prepare_inferior()
+
+        error = lldb.SBError()
+        content = self.process().ReadMemory(self.memory, 2 * self.pagesize, error)
+        self.assertEqual(len(content), self.pagesize)
+        self.assertEqual(content[0:7], b"needle\0")
+        self.assertTrue(error.Fail())
+        self.assertEqual(
+            f"memory read failed for {self.memory+self.pagesize:#x}", error.GetCString()
+        )
```
Comment (labath) [inline lldb/test/API/functionalities/memory/holes/TestMemoryHoles.py:63]: I guess this kind of error behavior can be useful in some situations (you get the memory that can be read *and* the error for the part that cannot), but I've found it very counterintuitive to use. I'd definitely be open to changing that (returning an error only if we didn't read anything).

File: lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationServerLLGS.cpp [hunk 2]
```diff
@@ -2526,28 +2526,16 @@ GDBRemoteCommunicationServerLLGS::Handle_memory_read(
...
       read_addr, &buf[0], byte_count, bytes_read);
-  if (error.Fail()) {
-    LLDB_LOGF(log,
-              "GDBRemoteCommunicationServerLLGS::%s pid %" PRIu64
-              " mem 0x%" PRIx64 ": failed to read. Error: %s",
-              __FUNCTION__, m_current_process->GetID(), read_addr,
-              error.AsCString());
+  LLDB_LOG(log, "ReadMemoryWithoutTrap({0}) read {1}/{2} bytes (error: {3})",
+           read_addr, byte_count, bytes_read, error);
+  if (bytes_read == 0)
```
Comment (labath) [inline lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationServerLLGS.cpp:2531]: This was changed to not assume that an error means nothing was read.

**Review by clayborg — APPROVED** [review]

File: lldb/source/Plugins/Process/Linux/NativeProcessLinux.cpp [hunk 3]
```diff
@@ -1641,6 +1641,10 @@ NativeProcessLinux::GetSoftwareBreakpointTrapOpcode(size_t size_hint) {
 
 Status NativeProcessLinux::ReadMemory(lldb::addr_t addr, void *buf, size_t size,
                                       size_t &bytes_read) {
+  Log *log = GetLog(POSIXLog::Memory);
+  LLDB_LOG(log, "addr = {0}, buf = {1}, size = {2}", addr, buf, size);
```
Comment (DavidSpickett) [inline lldb/source/Plugins/Process/Linux/NativeProcessLinux.cpp:1645]: Does this include the function name? If not, add something to say it's a read.

File: lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationServerLLGS.cpp (hunk 2 above)
Comment (DavidSpickett) [reply lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationServerLLGS.cpp:2531]: I would update the x/y bit to 1 be clearer which is which like:
```
requested X bytes read Y
```
(or use the phrasing from the log message you are removing below)

Also what is `error` here, `ReadMemoryWithoutTrap` doesn't seem to modify it.

File: lldb/test/API/functionalities/memory/holes/TestMemoryHoles.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        self.pagesize = self.frame().FindVariable("pagesize").GetValueAsUnsigned()
+        positions = self.frame().FindVariable("positions")
+        self.positions = [
+            positions.GetChildAtIndex(i).GetValueAsUnsigned()
+            for i in range(positions.GetNumChildren())
+        ]
+        self.assertEqual(len(self.positions), 5)
+
+    @expectedFailureWindows
+    def test_memory_read(self):
```
Comment (DavidSpickett) [inline lldb/test/API/functionalities/memory/holes/TestMemoryHoles.py:53]: To be clear, this test is not expecting memory read to skip over a hole. Merely handle there being a failure to read, correct?

Later changes (in your other PRs I think) will add memory find, which does have to know to jump over the holes.

File: lldb/test/API/functionalities/memory/holes/TestMemoryHoles.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
+"""
+Test the memory commands operating on memory regions with holes
```
Comment (DavidSpickett) [inline lldb/test/API/functionalities/memory/holes/TestMemoryHoles.py:2]: I don't mind the terminology but perhaps add ` with holes (inaccessible pages)` to help anyone who is skimming through this. So they only have to read the header to know whether to care about the rest of the file.

File: lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationServerLLGS.cpp (hunk 2 above)
Comment (labath) [reply lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationServerLLGS.cpp:2531]: Expanded the wording. I'm not sure what you meant with the second part. `error` is the result of the `ReadMemoryWithoutTrap` call (line 2527).

File: lldb/test/API/functionalities/memory/holes/TestMemoryHoles.py (hunk 5 above)
Comment (labath) [reply lldb/test/API/functionalities/memory/holes/TestMemoryHoles.py:2]: Done.

File: lldb/test/API/functionalities/memory/holes/TestMemoryHoles.py (hunk 4 above)
Comment (labath) [reply lldb/test/API/functionalities/memory/holes/TestMemoryHoles.py:53]: Correct.

File: lldb/test/API/functionalities/memory/holes/TestMemoryHoles.py (hunk 1 above)
Comment (labath) [reply lldb/test/API/functionalities/memory/holes/TestMemoryHoles.py:63]: (Removed the assertion on the exact error message, as windows returns a different error)

File: lldb/source/Plugins/Process/Linux/NativeProcessLinux.cpp (hunk 3 above)
Comment (labath) [reply lldb/source/Plugins/Process/Linux/NativeProcessLinux.cpp:1645]: it does (with `log enable -F`)

File: lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationServerLLGS.cpp (hunk 2 above)
Comment (DavidSpickett) [reply lldb/source/Plugins/Process/gdb-remote/GDBRemoteCommunicationServerLLGS.cpp:2533]: So it is :)

**Review by DavidSpickett — APPROVED** [review]
LGTM

**Pull request merged**